File: web/src/main/java/top/bettercode/summer/web/form/FormDuplicateCheckInterceptor.java
package top.bettercode.summer.web.form;

import org.springframework.web.method.HandlerMethod;
import top.bettercode.summer.tools.lang.util.AnnotatedUtils;
import top.bettercode.summer.web.servlet.NotErrorHandlerInterceptor;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * 表单重复检
 *
 * @author Peter Wu
 */
public class FormDuplicateCheckInterceptor implements NotErrorHandlerInterceptor {

    public static final String FORM_KEY = FormDuplicateCheckInterceptor.class.getName() + ".form_key";
    public final static String DEFAULT_MESSAGE = "您提交的太快了，请稍候再试。";
    private final IFormkeyService formkeyService;
    private final String formKeyName;

    public FormDuplicateCheckInterceptor(IFormkeyService formkeyService, String formKeyName) {
        this.formkeyService = formkeyService;
        this.formKeyName = formKeyName;
    }


    @Override
    public boolean preHandlerMethod(HttpServletRequest request, HttpServletResponse response,
                                    HandlerMethod handler) {
        FormDuplicateCheck annotation = AnnotatedUtils.getAnnotation(handler, FormDuplicateCheck.class);
        return formkeyService.checkRequest(request, formKeyName, annotation != null, annotation == null ? -1 : annotation.expireSeconds(), annotation == null ? DEFAULT_MESSAGE : annotation.message());
    }


    @Override
    public void afterCompletionMethod(HttpServletRequest request, HttpServletResponse response, HandlerMethod handler, Throwable ex) {
        if (ex == null) {
            ex = getError(request);
        }

        if (ex != null) {
            formkeyService.cleanKey(request);
        }
    }


}

File: web/src/main/kotlin/top/bettercode/summer/web/form/FormDuplicateCheckInterceptor.kt
package top.bettercode.summer.web.form

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.method.HandlerMethod
import top.bettercode.summer.tools.lang.util.AnnotatedUtils.getAnnotation
import top.bettercode.summer.web.servlet.NotErrorHandlerInterceptor
import java.time.Duration

/**
 * 表单重复检
 *
 * @author Peter Wu
 */
class FormDuplicateCheckInterceptor(
    private val formkeyService: IFormkeyService,
    private val formKeyName: String
) : NotErrorHandlerInterceptor {
    override fun preHandlerMethod(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: HandlerMethod
    ): Boolean {
        val annotation = getAnnotation(handler, FormDuplicateCheck::class.java)
        val expireSeconds = annotation?.expireSeconds
        val ttl = if (expireSeconds != null && expireSeconds > 0) {
            Duration.ofSeconds(expireSeconds)
        } else {
            null
        }
        formkeyService.duplicateCheck<Any>(
            request = request,
            formKeyName = formKeyName,
            autoFormKey = annotation != null,
            ttl = ttl,
            message = annotation?.message,
            ignoreHeaders = annotation?.ignoreHeaders,
            ignoreParams = annotation?.ignoreParams
        )
        return true
    }

    override fun afterCompletionMethod(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: HandlerMethod,
        ex: Throwable?
    ) {
        var e = ex
        if (e == null) {
            e = NotErrorHandlerInterceptor.getError(request)
        }
        if (e != null) {
            formkeyService.cleanKey(request)
        }
    }

    companion object {
        val FORM_KEY = FormDuplicateCheckInterceptor::class.java.name + ".form_key"
        const val DEFAULT_MESSAGE = "form.duplicate"
    }
}
